### ISSUE TITLE
define() with a namespace acts strange

### ISSUE THREAD

**Issue by bugreportuser** [vimeo/psalm#1301]
Constants defined with namespaces cause strange errors.

Example (https://getpsalm.org/r/6f2a74c453):
```php
<?php
namespace {
define('ns1\\cons1', 0);

echo \ns1\cons1; // false positive
echo ns1\cons1; // false positive
}
namespace ns2 {
define(__NAMESPACE__.'\\cons2', 0);

echo \ns2\cons2;
echo \ns3\cons2; // false negative
echo cons2;
}
namespace ns2 {
echo \ns2\cons2; // false positive
echo cons2; // false positive
}
namespace {
echo \ns2\cons2; // false positive
echo ns2\cons2; // false positive
}
```

Example output:
```
Psalm output (using commit 88b4f82):

ERROR: UndefinedConstant - 5:6 - Const ns1\cons1 is not defined

ERROR: UndefinedConstant - 6:6 - Const ns1\cons1 is not defined

ERROR: UndefinedConstant - 16:6 - Const ns2\cons2 is not defined

ERROR: UndefinedConstant - 17:6 - Const cons2 is not defined

ERROR: UndefinedConstant - 20:6 - Const ns2\cons2 is not defined

ERROR: UndefinedConstant - 21:6 - Const ns2\cons2 is not defined
```

**Comment by bugreportuser** [comment]
I now know what causes this. However, to send a pull request, I would need to know how Psalm resolves concatenation.
- The first constant is defined including the namespace but looked up without it. Fixing this is easy. The namespace can be removed when defining the constant.
    - When this issue is closed, I will open another about false negatives with constants that will be less important. `const` has a similar problem.
- The second constant is defined with a name that isn't a literal, so constants in the same block are assumed to exist. How would Psalm usually get the actual string?
    - Phan had the same problem: https://github.com/phan/phan/issues/2116

https://github.com/vimeo/psalm/blob/9a8f969d5ceae8d0a70460c563e9081f1c6dbb89/src/Psalm/Internal/Visitor/ReflectorVisitor.php#L571-L591
